lxdock/provisioners/puppet.py

```diff
@@ -0,0 +1,122 @@
+import logging
+import shlex
+import tempfile
+
+from voluptuous import Any, IsDir, IsFile, Required, Schema
+
+from ..network import get_ipv4_ip
+
+from .base import Provisioner
+
+logger = logging.getLogger(__name__)
+
+
+class PuppetProvisioner(Provisioner):
+    """ Allows to perform provisioning operations using Puppet. """
+
+    name = 'puppet'
+
+    guest_required_packages_arch = ['openssh', 'puppet', ]
+    guest_required_packages_centos = ['openssh-server', ]
+    guest_required_packages_debian = ['apt-utils', 'aptitude', 'openssh-server', 'puppet', ]
+    guest_required_packages_fedora = ['openssh-server', 'puppet', ]
+    guest_required_packages_ubuntu = ['apt-utils', 'aptitude', 'openssh-server', 'puppet', ]
+
+    # Refs Vagrant docs:
+    # https://www.vagrantup.com/docs/provisioning/puppet_apply.html#options
+    schema = {
+        Required('manifest_file'): IsFile(),
+        'environment': str,
+        'environment_path': IsDir(),
+        'module_path': IsDir(),
+        'facter': dict,
+        'options': str,
+    }
+
+    _guest_manifest_file = '/tmp/site.pp'
+    _guest_module_path = '/tmp/puppet-modules'
+    _guest_environment_path = '/tmp/puppet-environment'
+
+    def provision(self):
+        """ Performs the provisioning operations using puppet. """
+        ip = get_ipv4_ip(self.guest.lxd_container)
+
```

Comment:
Note to myself: check if `puppet` exists before proceeding. If it doesn't exist, raise an error to let user know.